### PR TITLE
Provide general design guidelines per response-of-T discussion.

### PR DESCRIPTION
**Pull request by johanste**

### CODE REVIEW

File: docs/design/ResponseFormats.mdk [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+~ Must {general-return-logical-entities}
+optimize for returning the logical entity for a given request. The logical entity MUST represent the information needed in the 99%+ case.
+~
+
+~ Must {general-return-expose-raw}
+*make it possible* for a developer to get access to the complete response, including the status line, headers and body. The client library MUST follow the language specific guidance for accomplishing this.
+~
+
+~ Must {general-return-expose-stream}
+make it possible to access the body of the response as a stream.
```
Comment (JeffreyRichter) [inline docs/design/ResponseFormats.mdk]: Blob/file download needs this. But for all other "payload" responses, I would think that the response stream is read and deserialized and hence, the stream is now empty (any probably even closed).

Comment (johanste) [reply docs/design/ResponseFormats.mdk]: I assume this depends on when the actual deserialization happens. For Python, the caller can specify if they want the response to be returned as a stream (in which case deserialization doesn't happen). And I believe that the current C# pipeline/returning Response<T> deserializes when the consumer accesses T (the Result() method on the Response<T> instance). But I may be wrong there.

I know that there are other services that have expressed a desire to support this. Consider Cosmos where a client is trying to download what is effectively an array of json documents and store the result on disk. De/re serializing the payload is costly in that scenario.

Are you suggesting that we don't need this as a general guidance and only support it on a case-by-case basis?

**Comment by JeffreyRichter** [comment]
I believe that (except for blob/file downloads or similar) the response should be read from the service in it's entirety between the retry and transport policies because the connection could close while reading and then retry doesn't work any more.

If this is done, then the steam will be empty by the time you hit the top of the pipeline.


  *   Jeffrey Richter (watch Architecting Distributed Cloud Apps<https://azure.microsoft.com/training/distributed-cloud-apps/>)

________________________________

File: docs/design/ResponseFormats.mdk [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+~ Must {general-return-expose-stream}
+make it possible to access the body of the response as a stream. 
+~
+
+~ Must {general-return-document-raw-stream}
+document and provide examples on how to access the raw and streamed response for a given request.
+~
+
+~ Must {general-return-pageable-iterate-items}
+provide a, for the language idiomatic, way to enumerate all logical entities for a paged operation, automatically fetching new pages as needed.
```
Comment (JonathanGiles) [inline docs/design/ResponseFormats.mdk]:
```suggestion
provide a language idiomatic way to enumerate all logical entities for a paged operation, automatically fetching new pages as needed.
```

File: docs/design/ResponseFormats.mdk (hunk 1 above)
Comment (johanste) [reply docs/design/ResponseFormats.mdk]: I've removed the requirement to expose the response as a stream.

File: docs/design/ResponseFormats.mdk [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+For methods that makes a single logical request, or a deterministic sequence of requests:
+
+An example of a *single logical request* is a request that may be retried inside the operation.
+
+An example of a *deterministic sequence of requests* is a paged operation.
+
+The *logical entity* is a protocol neutral representation of a response. For HTTP, the logical entity may combine data from headers, body and the status line. A common example is exposing an ETag header as a property on the logical entity in addition to any deserialized content from the body.
+
+~ Must {general-return-logical-entities}
```
Comment (adrianhall) [inline docs/design/ResponseFormats.mdk]: I think you are missing hashes from inside the braces.

File: docs/design/ResponseFormats.mdk [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+*make it possible* for a developer to get access to the complete response, including the status line, headers and body. The client library MUST follow the language specific guidance for accomplishing this.
+~
+
+~ Must {general-return-document-raw-stream}
+document and provide examples on how to access the raw and streamed response for a given request.
+~
+
+~ Must {general-return-pageable-iterate-items}
+provide a language idiomatic way to enumerate all logical entities for a paged operation, automatically fetching new pages as needed.
+
```
Comment (adrianhall) [inline docs/design/ResponseFormats.mdk]: Now that paging is approved, perhaps include a link to the paging guidance?

Comment (johanste) [reply docs/design/ResponseFormats.mdk]: Should we merge the two documents?

Comment (adrianhall) [reply docs/design/ResponseFormats.mdk]: Could be a reasonable solution.  I'll leave that up to yourself and @bterlson

Comment (johanste) [reply docs/design/ResponseFormats.mdk]: I removed the reference to paging from this document now.

File: docs/design/ResponseFormats.mdk (hunk 3 above)
Comment (johanste) [reply docs/design/ResponseFormats.mdk]: Fixed.

**Pull request merged**